middlewares/authorizeRoles.js

```diff
@@ -0,0 +1,28 @@
+const { VALID_ROLES } = require("../constants/roles");
+
+/**
+ * Create an authorization middleware for a route based on the required role needed
+ * for that route.
+ * Note: This must be added on routes after the `authenticate` middleware.
+ * @param {Array.<String>} allowedRoles - Roles allowed for a route.
+ * @returns {Function} - A middleware function that authorizes given role.
+ */
+const authorizeRoles = (allowedRoles) => {
+  return (req, res, next) => {
+    const { roles = {} } = req.userData;
+
+    const rolesAreValid = allowedRoles.every((role) => Object.values(VALID_ROLES).includes(role));
+    if (!rolesAreValid) {
+      return res.boom.badImplementation("Route authorization failed. Please contact admin");
+    }
+
+    const userHasPermission = allowedRoles.some((role) => roles[`${role}`]);
```

Comment:
How will the system handle cases where certain roles are blocklisted for the user.

e.g Maybe a particular 'member' is also added in the 'timeout' role, where they should not be able to use certain features.
Since the 'memebers' role will match, this user will be able to continue being authorized.

Reply:
Also - https://martinfowler.com/bliki/Yagni.html